### PR TITLE
refactor resource logic

### PR DESCRIPTION
**Pull request by xi**
This implements a first part of #2206: Most functionality from resource classes is moved to `ResourceUtil.ts`. I also tried to make the usage of types a bit more consistent.

While working on this I think I found a bug in `deepPost()` (see 1ca8583). I want to check with some backend people before merging this.

### CODE REVIEW

**Comment by xi** [comment]
> While working on this I think I found a bug in deepPost() (see 1ca8583). I want to check with some backend people before merging this.

I opened #2211 for that.

File: src/adhocracy_frontend/adhocracy_frontend/static/js/ResourcesBase.ts [hunk 1]
```diff
@@ -55,16 +53,4 @@ export class Resource implements IResource {
         this.data = {};
     }
     /* tslint:enable:variable-name */
-
-    public isInstanceOf(resourceType : string) : boolean {
-        var _class : IResourceClass = <any>this.constructor;
```
Comment (Johnschmidfront) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/ResourcesBase.ts]: (For my better understanding:) In general, this `IResourceClass` has been generated in MkResources.ts, right?

File: src/adhocracy_frontend/adhocracy_frontend/static/js/ResourcesBase.ts [hunk 2]
```diff
@@ -3,9 +3,7 @@ export interface ISheetMetaApi {
     readable : string[];
     editable : string[];
     creatable : string[];
-    /* tslint:disable:variable-name */
```
Comment (Johnschmidfront) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/ResourcesBase.ts]: Are these changes related to 6620eeac45ea3f0c299643e8df20a4824f3c45d7 (mv `isInstanceOf`to ResourceUtil)?

File: src/adhocracy_frontend/adhocracy_frontend/static/js/ResourcesBase.ts [hunk 3]
```diff
@@ -25,9 +19,9 @@ export interface IResource {
     data : Object;
     path : string;
     content_type : string;
-    parent : string;
-    first_version_path : string;
-    root_versions : string[];
+    parent? : string;
```
Comment (Johnschmidfront) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/ResourcesBase.ts]: How come this?

File: src/adhocracy_frontend/adhocracy_frontend/static/js/ResourcesBase.ts (hunk 2 above)
Comment (xi) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/ResourcesBase.ts]: It is simply a tslint exception. It is not required here so I removed it.

File: src/adhocracy_frontend/adhocracy_frontend/static/js/ResourcesBase.ts (hunk 1 above)
Comment (xi) [inline src/adhocracy_frontend/adhocracy_frontend/static/js/ResourcesBase.ts]: `IRresourceClass` is defined above. The concrete class is generated by `mkResources.ts`.

**Comment by Johnschmidfront** [comment]
looks good & works as far as I can tell.

**Comment by Johnschmidfront** [comment]
Unfortunately, my local acceptance tests seem to be broken. Can somebody else run them please?

**Comment by Johnschmidfront** [comment]
Now, my acceptance tests are rational again. 5 tests have failed (that do not fail on `master`):
 - comments can be edited
 - comments can be edited twice
 - comments can be edited after child edit
 - comments can be edited and replied
 - user password reset error displayed if the email is not associated to an user

**Comment by xi** [comment]
We are about to release a new version of a3. This pull request is on hold until after that.

**Comment by xi** [comment]
Most tests seem to be related to comment editing. When I tried that locally I got the following error message:

   unknown field named preliminaryNames in sheet adhocracy_core.sheets.rate.IRateable

This message is thrown by `adhMetaApi.field()` if the requested field does not exist. The issue should be fixed with 7fa8fb6.

**Comment by Johnschmidfront** [comment]
still looks good.

**Comment by Johnschmidfront** [comment]
my acceptance tests now pass.

**Pull request closed without merging**